File: news_server/modules/source/infrastructure/getters/bilibili.ts
/**
 * 哔哩哔哩新闻源获取器
 * 从 newsnow/server/sources/bilibili.ts 迁移
 */
import type { SourceGetter } from '../../domain/getter/getter'
import { NewsItem } from '../../domain/newsitem/entity'

interface BilibiliHotSearchResponse {
  code: number
  list: {
    keyword: string
    show_name: string
    icon?: string
  }[]
}

interface BilibiliHotVideoResponse {
  code: number
  data: {
    list: {
      bvid: string
      title: string
      pic: string
      owner: {
        name: string
      }
      stat: {
        view: number
        like: number
      }
      desc: string
      pubdate: number
    }[]
  }
}

function formatNumber(num: number): string {
  if (num >= 10000) {
    return `${Math.floor(num / 10000)}w+`
  }
  return num.toString()
}

export const bilibiliHotSearchGetter: SourceGetter = async () => {
  const url = 'https://s.search.bilibili.com/main/hotword?limit=30'
  const response = await fetch(url)
  
  if (!response.ok) {
    throw new Error(`Failed to fetch Bilibili hot search: ${response.statusText}`)
  }

  const res = (await response.json()) as BilibiliHotSearchResponse

  return res.list.map((item) =>
    NewsItem.fromEntity({
      id: item.keyword,
      title: item.show_name,
      url: `https://search.bilibili.com/all?keyword=${encodeURIComponent(item.keyword)}`,
      extra: {
        icon: item.icon || false,
      },
    })
  )
}

export const bilibiliHotVideoGetter: SourceGetter = async () => {
  const url = 'https://api.bilibili.com/x/web-interface/popular'
  const response = await fetch(url, {
    headers: {
      'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36',
    },
  })
  
  if (!response.ok) {
    throw new Error(`Failed to fetch Bilibili hot video: ${response.statusText}`)
  }

  const res = (await response.json()) as BilibiliHotVideoResponse

  if (res.code !== 0 || !res.data || !res.data.list) {
    throw new Error(`Bilibili API error: ${res.code}`)
  }

  return res.data.list.map((video) =>
    NewsItem.fromEntity({
      id: video.bvid,
      title: video.title,
      url: `https://www.bilibili.com/video/${video.bvid}`,
      pubDate: video.pubdate * 1000,
      extra: {
        info: `${video.owner.name} · ${formatNumber(video.stat.view)}观看 · ${formatNumber(video.stat.like)}点赞`,
        hover: video.desc,
        icon: video.pic
          ? {
              url: video.pic,
              scale: 1,
            }
          : undefined,
      },
    })
  )
}

export const bilibiliRankingGetter: SourceGetter = async () => {
  const url = 'https://api.bilibili.com/x/web-interface/ranking/v2'
  const response = await fetch(url, {
    headers: {
      'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36',
    },
  })
  
  if (!response.ok) {
    throw new Error(`Failed to fetch Bilibili ranking: ${response.statusText}`)
  }

  const res = (await response.json()) as BilibiliHotVideoResponse

  if (res.code !== 0 || !res.data || !res.data.list) {
    throw new Error(`Bilibili API error: ${res.code}`)
  }

  return res.data.list.map((video) =>
    NewsItem.fromEntity({
      id: video.bvid,
      title: video.title,
      url: `https://www.bilibili.com/video/${video.bvid}`,
      pubDate: video.pubdate * 1000,
      extra: {
        info: `${video.owner.name} · ${formatNumber(video.stat.view)}观看 · ${formatNumber(video.stat.like)}点赞`,
        hover: video.desc,
        icon: video.pic
          ? {
              url: video.pic,
              scale: 1,
            }
          : undefined,
      },
    })
  )
}

File: modules/source/infrastructure/getters/bilibili.go
package getters

import (
	"context"
	"encoding/json"
	"fmt"
	"net/url"
)

func formatNumber(num int) string {
	if num >= 10000 {
		return fmt.Sprintf("%dw+", num/10000)
	}
	return fmt.Sprintf("%d", num)
}

func (r *Registry) bilibiliHot(ctx context.Context) ([]Item, error) {
	body, err := r.get(ctx, "https://s.search.bilibili.com/main/hotword?limit=30", nil)
	if err != nil {
		return nil, err
	}
	var res struct {
		List []struct {
			Keyword  string `json:"keyword"`
			ShowName string `json:"show_name"`
			Icon     string `json:"icon"`
		} `json:"list"`
	}
	if err := json.Unmarshal(body, &res); err != nil {
		return nil, err
	}
	out := make([]Item, 0, len(res.List))
	for _, k := range res.List {
		out = append(out, Item{
			ID: k.Keyword, Title: k.ShowName,
			URL:   "https://search.bilibili.com/all?keyword=" + url.QueryEscape(k.Keyword),
			Extra: map[string]any{"icon": k.Icon},
		})
	}
	return out, nil
}

func (r *Registry) bilibiliVideo(ctx context.Context) ([]Item, error) {
	return r.bilibiliList(ctx, "https://api.bilibili.com/x/web-interface/popular")
}

func (r *Registry) bilibiliRank(ctx context.Context) ([]Item, error) {
	return r.bilibiliList(ctx, "https://api.bilibili.com/x/web-interface/ranking/v2")
}

func (r *Registry) bilibiliList(ctx context.Context, api string) ([]Item, error) {
	body, err := r.get(ctx, api, nil)
	if err != nil {
		return nil, err
	}
	var res struct {
		Code int `json:"code"`
		Data struct {
			List []struct {
				BVID  string `json:"bvid"`
				Title string `json:"title"`
				Pic   string `json:"pic"`
				Desc  string `json:"desc"`
				Owner struct {
					Name string `json:"name"`
				} `json:"owner"`
				Stat struct {
					View int `json:"view"`
					Like int `json:"like"`
				} `json:"stat"`
				Pubdate int64 `json:"pubdate"`
			} `json:"list"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &res); err != nil {
		return nil, err
	}
	if res.Code != 0 {
		return nil, fmt.Errorf("bilibili API error: %d", res.Code)
	}
	out := make([]Item, 0, len(res.Data.List))
	for _, v := range res.Data.List {
		extra := map[string]any{
			"info":  fmt.Sprintf("%s · %s观看 · %s点赞", v.Owner.Name, formatNumber(v.Stat.View), formatNumber(v.Stat.Like)),
			"hover": v.Desc,
		}
		if v.Pic != "" {
			extra["icon"] = map[string]any{"url": v.Pic, "scale": 1}
		}
		out = append(out, Item{ID: v.BVID, Title: v.Title, URL: "https://www.bilibili.com/video/" + v.BVID, PubDate: v.Pubdate * 1000, Extra: extra})
	}
	return out, nil
}
